synapse/util/async_helpers.py

```diff
@@ -234,6 +236,59 @@ def yieldable_gather_results(
     ).addErrback(unwrapFirstError)
 
 
+T1 = TypeVar("T1")
+T2 = TypeVar("T2")
+T3 = TypeVar("T3")
+
+
+@overload
+def gather_results(
+    deferredList: Tuple[()], consumeErrors: bool = ...
+) -> "defer.Deferred[Tuple[()]]":
+    ...
+
+
+@overload
+def gather_results(
+    deferredList: Tuple["defer.Deferred[T1]"],
+    consumeErrors: bool = ...,
+) -> "defer.Deferred[Tuple[T1]]":
+    ...
+
+
+@overload
+def gather_results(
+    deferredList: Tuple["defer.Deferred[T1]", "defer.Deferred[T2]"],
+    consumeErrors: bool = ...,
+) -> "defer.Deferred[Tuple[T1, T2]]":
+    ...
+
+
+@overload
+def gather_results(
+    deferredList: Tuple[
+        "defer.Deferred[T1]", "defer.Deferred[T2]", "defer.Deferred[T3]"
+    ],
+    consumeErrors: bool = ...,
+) -> "defer.Deferred[Tuple[T1, T2, T3]]":
+    ...
+
+
+def gather_results(  # type: ignore[misc]
+    deferredList: Tuple["defer.Deferred[T1]", ...],
+    consumeErrors: bool = False,
+) -> "defer.Deferred[Tuple[T1, ...]]":
```

Comment:
I'm getting C++ [parameter pack](https://en.cppreference.com/w/cpp/language/parameter_pack) vibes here!

I guess the overloads do the trick.

Reply:
If anyone asks, I'm going to claim that the overloads were inspired by [Rust](https://doc.rust-lang.org/std/primitive.tuple.html#impl-Debug-4) rather than [C#](https://docs.microsoft.com/en-us/dotnet/api/system.valuetuple-4) or [TypeScript](https://github.com/microsoft/TypeScript/blob/4e39023396f03907f7c9fecb9ad72d5f42256eff/lib/lib.es2015.promise.d.ts#L41).